go/vaas/pkg/vaas/vaas.go

```diff
@@ -0,0 +1,323 @@
+package vaas
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	BroadcastChannel "vaas/pkg/broadcast_channel"
+	"vaas/pkg/hash"
+	msg "vaas/pkg/messages"
+	"vaas/pkg/options"
+
+	"github.com/gorilla/websocket"
+)
+
+const TIMEOUT = 60
+
+type Vaas struct {
+	sessionId           string
+	websocketConnection *websocket.Conn
+	broadcastChannel    *BroadcastChannel.BroadcastChannel[msg.VerdictResponse]
+	requestChannel      chan msg.IVerdictRequest
+	responseChannel     chan msg.VerdictResponse
+	Url                 url.URL `default:"wss://gateway-vaas.gdatasecurity.de"`
+	options             options.VaasOptions
+	Ctx                 context.Context
+}
+
+func New(options options.VaasOptions) *Vaas {
+	rc := make(chan msg.VerdictResponse)
+	ctx := context.Background()
+	bc := BroadcastChannel.New(ctx, rc)
+
+	vaas := &Vaas{
+		options:          options,
+		requestChannel:   make(chan msg.IVerdictRequest, 1),
+		responseChannel:  rc,
+		broadcastChannel: bc,
+		Ctx:              ctx,
+	}
+
+	return vaas
+}
+
+func (v *Vaas) Connect(token string) error {
+	connection, _, websocketErr := websocket.DefaultDialer.Dial("wss://gateway-vaas.gdatasecurity.de", nil)
+	if websocketErr != nil {
+		return websocketErr
+	}
+	v.websocketConnection = connection
+
+	if err := v.Authenticate(token); err != nil {
+		return errors.New("failed to authenticate: " + err.Error())
+	}
+
+	go v.sendRequests()
+	go v.readResponses()
+
+	return nil
+}
+
+func (v *Vaas) Authenticate(token string) error {
+	v.websocketConnection.WriteJSON(msg.AuthRequest{
+		Kind:  "AuthRequest",
+		Token: token,
+	})
+
+	var authResponse msg.AuthResponse
+	v.websocketConnection.ReadJSON(&authResponse)
+	if authResponse.Kind == "Error" {
+		return errors.New(authResponse.Text)
+	}
+	if !authResponse.Success {
+		return errors.New("failed to authenticate")
+	}
+
+	v.sessionId = authResponse.SessionId
+
+	return nil
+}
+
+func (v *Vaas) ForSha256(sha256 string) (msg.VaasVerdict, error) {
+	if v.sessionId == "" {
+		return msg.VaasVerdict{}, errors.New("invalid operation")
+	}
+	subscription := v.broadcastChannel.Subscribe()
+	defer v.broadcastChannel.RemoveSubscription(subscription)
+	request := msg.VerdictRequest{}.New(v.sessionId, v.options, sha256)
+	v.requestChannel <- request
+
+	verdictResponse, responseErr := v.waitForResponse(subscription, request.Guid)
+	if responseErr != nil {
+		return msg.VaasVerdict{
+			Verdict: msg.Verdict(msg.Error),
+			Sha256:  sha256,
+		}, responseErr
+	}
+
+	return msg.VaasVerdict{
+		Verdict: verdictResponse.Verdict,
+		Sha256:  verdictResponse.Sha256,
+	}, nil
+}
+
+func (v *Vaas) ForSha256List(sha256List []string) ([]msg.VaasVerdict, error) {
+	if v.sessionId == "" {
+		return []msg.VaasVerdict{}, errors.New("invalid operation")
```

Comment:
If you use the same error multiple time, you should define it globally for the package:
```go
var ErrInvalidOperation = errors.New("invalid operation)
```
This way the caller has the chance to check for specific errors as well:
```go
if errors.Is(err, ErrInvalidOperation) {
    // do some stuff
}
```